draft-ietf-tls-esni.md

```diff
@@ -948,11 +948,17 @@ messages between the client and backend server unmodified.
 
 Otherwise, if all candidate ECHConfig values fail to decrypt the extension, the
 client-facing server MUST ignore the extension and proceed with the connection
-using ClientHelloOuter. This connection proceeds as usual, except the server
-MUST include the "encrypted_client_hello" extension in its EncryptedExtensions
-with the "retry_configs" field set to one or more ECHConfig structures with
-up-to-date keys. Servers MAY supply multiple ECHConfig values of different
-versions. This allows a server to support multiple versions at once.
+using ClientHelloOuter. If this does not result in sending a HelloRetryRequest,
+the connection proceeds as usual, except the server MUST include the
+"encrypted_client_hello" extension in its EncryptedExtensions with the
+"retry_configs" field set to one or more ECHConfig structures with up-to-date
+keys. Servers MAY supply multiple ECHConfig values of different versions. This
+allows a server to support multiple versions at once. Otherwise, if this results
+in sending a HelloRetryRequest, and the corresponding ClientHello included an
+"encrypted_client_hello" extension, the client-facing server MAY include the
+"encrypted_client_hello" extension in the HelloRetryRequest with a payload of
+8 random bytes; see {{dont-stick-out}} for details. Afterwards, the
+client-facing server proceeds as described in {{client-facing-server-hrr}}.
```

Comment:
This is a little long. How about:

```
Otherwise, if all candidate ECHConfig values fail to decrypt the extension, the
client-facing server MUST ignore the extension and proceed with the connection
using ClientHelloOuter, with the following modifications:

* If sending a HelloRetryRequest, the server MAY include an
  "encrypted_client_hello" extension with a payload of 8 random bytes; see
  {{dont-stick-out}} for details.

* The server MUST include the "encrypted_client_hello" extension in its
  EncryptedExtensions message with ....
```

This also fixes the issue where we need to duplicate the EncryptedExtensions text (see comment below).

Though, it occurs to me, even that's a little odd because even non-ECH servers could send HRR.ECH for GREASE. Maybe we can handle that be deleting a few words. Suppose you are a server with zero ECHConfigs. The dispatch in {#server-behavior} still sends you to this section. The candidate ECHConfig values is then empty, so you'll hit this case. That means we actually should replace:

> The server MUST include the "encrypted_client_hello" extension in its EncryptedExtensions message with [...]

with:

> If the server is configured with any ECHConfigs, it MUST include the "encrypted_client_hello" extension in its EncryptedExtensions message with [...]

Reply:
That modification is good in both places -- I'll take it!